core/src/main/java/org/kiwix/kiwixmobile/core/main/MainPresenter.kt

```diff
@@ -0,0 +1,142 @@
+import io.reactivex.CompletableObserver
+import io.reactivex.SingleObserver
+import io.reactivex.disposables.Disposable
+import org.kiwix.kiwixmobile.core.base.BasePresenter
+import org.kiwix.kiwixmobile.core.bookmark.BookmarkItem
+import org.kiwix.kiwixmobile.core.data.DataSource
+import org.kiwix.kiwixmobile.core.di.ActivityScope
+import org.kiwix.kiwixmobile.core.history.HistoryListItem.HistoryItem
+import org.kiwix.kiwixmobile.core.main.MainContract.Presenter
+import org.kiwix.kiwixmobile.core.main.MainContract.View
+import org.kiwix.kiwixmobile.core.zim_manager.fileselect_view.adapter.BooksOnDiskListItem
+import org.kiwix.kiwixmobile.core.zim_manager.fileselect_view.adapter.BooksOnDiskListItem.BookOnDisk
+import javax.inject.Inject
+
+@ActivityScope
+internal class MainPresenter @Inject constructor(private val dataSource: DataSource) :
+  BasePresenter<View?>(), Presenter {
+  override fun loadBooks() {
+    dataSource.languageCategorizedBooks
+      .subscribe(object : SingleObserver<List<BooksOnDiskListItem?>?> {
+        override fun onSubscribe(d: Disposable) {
+          compositeDisposable.add(d)
+        }
+
+        override fun onSuccess(books: List<BooksOnDiskListItem?>) {
+          view!!.addBooks(books)
+        }
+
+        override fun onError(e: Throwable) {
+          Log.e(
+            TAG,
+            "Unable to load books",
+            e
+          )
+        }
+      })
+  }
+
+  override fun saveBooks(book: List<BookOnDisk>) {
+    dataSource.saveBooks(book)
+      .subscribe(object : CompletableObserver {
+        override fun onSubscribe(d: Disposable) {
+          // TODO
+        }
+
+        override fun onComplete() {
+          loadBooks()
+        }
+
+        override fun onError(e: Throwable) {
+          Log.e(
+            TAG,
+            "Unable to save books",
+            e
+          )
+        }
+      })
+  }
+
+  override fun saveHistory(history: HistoryItem) {
+    dataSource.saveHistory(history)
+      .subscribe(object : CompletableObserver {
+        override fun onSubscribe(d: Disposable) {
+          // TODO
+        }
+
+        override fun onComplete() {
+          // TODO
+        }
+
+        override fun onError(e: Throwable) {
+          Log.e(TAG, "Unable to save history", e)
+        }
+      })
+  }
+
+  override fun saveBookmark(bookmark: BookmarkItem) {
+    dataSource.saveBookmark(bookmark)
+      .subscribe(object : CompletableObserver {
+        override fun onSubscribe(d: Disposable) {
+          // TODO
+        }
+
+        override fun onComplete() {
+          // TODO
+        }
+
+        override fun onError(e: Throwable) {
+          Log.e(
+            TAG,
+            "Unable to save bookmark",
+            e
+          )
+        }
+      })
+  }
+
+  override fun deleteBookmark(bookmarkUrl: String) {
+    dataSource.deleteBookmark(bookmarkUrl)
+      .subscribe(object : CompletableObserver {
+        override fun onSubscribe(d: Disposable) {
+          // TODO
```

Comment:
Why todo? This must be completely functional. All of these can be made concise as suggested by Sean.

Reply:
I am making the suggested changes by @macgills but the above "todo" was to prevent failing of "detekt".